Truncate error list so SNS message is less than 256kb

# Overview
We see validation jobs succeed, but hang in the tracker UI because Tracker receives no SNS message from the batch. This is because of the following error sending the SNS message:

2025-10-17 04:19:14 [ERROR] job_id=8e655fcc-3122-4d14-b40f-c67c6dd645f7 job_name=dataset-validator-9a7b2ba4-d109-4a94-a911-0f5369bece45 __main__: SNS publish failed [InvalidParameter]: Invalid parameter: Message too long

Let's truncate the message and add...
"Additional errors truncated" (or similar) to the end of the error list so it will show up in the tracker ui.

--- a/services/dataset-validator/src/dataset_validator/main.py
+++ b/services/dataset-validator/src/dataset_validator/main.py
@@ -133,7 +133,7 @@ class ValidationMessage:
         if self.tool_reports is None or len(full_message) < max_length:
             return full_message
         
-        # Create a list of message arrays from tool_reports, represented as (key in rool_reports, attribute in tool report object) pairs
+        # Create a list of message arrays from tool_reports, represented as (key in tool_reports, attribute in tool report object) pairs
         list_paths = [(key, message_type) for key in self.tool_reports.keys() for message_type in ["errors", "warnings"]]
         # Sort from longest to shortest list so that we won't remove shorter lists entirely before getting to a disproportionally long list that has to be truncated
         list_paths.sort(key=lambda p: self._json_length_of_report_list(*p), reverse=True)

--- a/services/dataset-validator/src/dataset_validator/main.py
+++ b/services/dataset-validator/src/dataset_validator/main.py
@@ -149,7 +149,7 @@ class ValidationMessage:
                 threshold_path = p
                 break
         
-        # If the message is somehow still too big, try using the truncated message (which will have all lists truncated)
+        # If all lists have been truncated and somehow none have made the message small enough, give up and try using it anyway
         if threshold_path is None:
             return truncated_message.to_json()
         

--- a/services/dataset-validator/src/dataset_validator/main.py
+++ b/services/dataset-validator/src/dataset_validator/main.py
@@ -4,6 +4,7 @@ Dataset Validator - AWS Batch Job
 Main entry point for dataset validation processing.
 """
 
+import copy
 import hashlib
 import json
 import logging
@@ -22,6 +23,10 @@ import anndata
 import pandas as pd
 from cap_upload_validator import UploadValidator
 from cap_upload_validator.errors import CapException, CapMultiException
+
+
+MAX_SNS_MESSAGE_LENGTH = 250_000 # Somewhat less than the actual value of 256KiB, to make room for small unforseen deviations
+TRUNCATED_MESSAGES_MESSAGE = "Messages truncated"
 
 
 # Environment variable constants
@@ -108,8 +113,62 @@ class ValidationMessage:
     error_message: Optional[str] = None                  # Human-readable error description
 
     def to_json(self) -> str:
-        """Convert to JSON string for SNS publishing."""
+        """Convert to JSON string."""
         return json.dumps(asdict(self), indent=2)
+    
+    def _get_report_message_list(self, report_key: str, message_type: str) -> List[str]:
+        return getattr(self.tool_reports[report_key], message_type)
+
+    def _set_report_message_list(self, report_key: str, message_type: str, value: List[str]):
+        setattr(self.tool_reports[report_key], message_type, value)
+
+    def _json_length_of_report_list(self, report_key: str, message_type: str) -> int:
+        return len(json.dumps(self._get_report_message_list(report_key, message_type)))
+
+    def to_length_limited_json(self, max_length=MAX_SNS_MESSAGE_LENGTH) -> str:
+        """Convert to a JSON string with message lists truncated to ensure that the JSON is below a given length."""
+
+        # If there are no reports to truncate or the full message is short enough, use the full message
+        full_message = self.to_json()
+        if self.tool_reports is None or len(full_message) < max_length:
+            return full_message
+        
+        # Create a list of message arrays from tool_reports, represented as (key in rool_reports, attribute in tool report object) pairs
+        list_paths = [(key, message_type) for key in self.tool_reports.keys() for message_type in ["errors", "warnings"]]
+        # Sort from longest to shortest list so that we won't remove shorter lists entirely before getting to a disproportionally long list that has to be truncated
+        list_paths.sort(key=lambda p: self._json_length_of_report_list(*p), reverse=True)
+
+        # Create a copy of the validation message to truncate lists in
+        truncated_message = copy.deepcopy(self)
+
+        # Truncate entire lists until the message JSON is small enough, and note which list was the last to be truncated
+        threshold_path = None
+        for p in list_paths:
+            truncated_message._set_report_message_list(*p, [TRUNCATED_MESSAGES_MESSAGE])
+            if len(truncated_message.to_json()) < max_length:
+                threshold_path = p
+                break
+        
+        # If the message is somehow still too big, try using the truncated message (which will have all lists truncated)
+        if threshold_path is None:
+            return truncated_message.to_json()
+        
+        # Get the original value of the list that was the threshold for making the message small enough
+        message_list = self._get_report_message_list(*threshold_path)
+
+        # Do a binary search to determine how much of the list can be retained
+        max_valid_length = 0
+        min_invalid_length = len(message_list)
+        while min_invalid_length - max_valid_length > 1:
+            middle_length = int((max_valid_length + min_invalid_length)/2)
+            truncated_message._set_report_message_list(*threshold_path, [*message_list[:middle_length], TRUNCATED_MESSAGES_MESSAGE])
+            if len(truncated_message.to_json()) < max_length:
+                max_valid_length = middle_length
+            else:
+                min_invalid_length = middle_length
+        
+        # Return the truncated message JSON
+        return truncated_message.to_json()
 
 
 def configure_logging() -> logging.Logger:
@@ -165,7 +224,7 @@ def publish_validation_result(message: ValidationMessage, sns_topic_arn: str) ->
         # Publish message
         response = sns_client.publish(
             TopicArn=sns_topic_arn,
-            Message=message.to_json(),
+            Message=message.to_length_limited_json(),
             Subject=f"Dataset Validation Result - {message.status.upper()}"
         )
         

--- a/services/dataset-validator/src/dataset_validator/main.py
+++ b/services/dataset-validator/src/dataset_validator/main.py
@@ -114,7 +114,7 @@ class ValidationMessage:
 
     def to_json(self) -> str:
         """Convert to JSON string."""
-        return json.dumps(asdict(self), indent=2)
+        return json.dumps(asdict(self), separators=(',', ':'))
     
     def _get_report_message_list(self, report_key: str, message_type: str) -> List[str]:
         return getattr(self.tool_reports[report_key], message_type)

--- a/services/dataset-validator/src/dataset_validator/main.py
+++ b/services/dataset-validator/src/dataset_validator/main.py
@@ -156,16 +156,18 @@ class ValidationMessage:
         # Get the original value of the list that was the threshold for making the message small enough
         message_list = self._get_report_message_list(*threshold_path)
 
-        # Do a binary search to determine how much of the list can be retained
+        # Do a binary search to determine how much of the list can be retained, ensuring that the truncated message remains at a valid length
         max_valid_length = 0
         min_invalid_length = len(message_list)
         while min_invalid_length - max_valid_length > 1:
+            truncated_list_before = truncated_message._get_report_message_list(*threshold_path)
             middle_length = int((max_valid_length + min_invalid_length)/2)
             truncated_message._set_report_message_list(*threshold_path, [*message_list[:middle_length], TRUNCATED_MESSAGES_MESSAGE])
             if len(truncated_message.to_json()) < max_length:
                 max_valid_length = middle_length
             else:
                 min_invalid_length = middle_length
+                truncated_message._set_report_message_list(*threshold_path, truncated_list_before)
         
         # Return the truncated message JSON
         return truncated_message.to_json()

--- a/services/dataset-validator/tests/test_validation_message.py
+++ b/services/dataset-validator/tests/test_validation_message.py
@@ -1,0 +1,89 @@
+import sys
+from pathlib import Path
+
+import pytest
+
+sys.path.insert(0, str(Path(__file__).parent.parent / "src"))
+from dataset_validator.main import ValidationToolReport, ValidationMessage
+
+@pytest.mark.parametrize("test_case", [
+    {
+        "name": "none_truncated",
+        "description": "Test length-limited JSON with no truncation needed",
+        "message_length": 100,
+        "message_counts": {
+          "cap": {
+            "errors": 50,
+            "warnings": 20
+          },
+          "cellxgene": {
+            "errors": 80,
+            "warnings": 30
+          }
+        },
+        "expected_min_length": 0,
+        "expected_max_length": 20000,
+        "expected_truncated_count": 0
+    },
+    {
+        "name": "one_truncated",
+        "description": "Test JSON with one message list requiring truncation",
+        "message_length": 100,
+        "message_counts": {
+          "cap": {
+            "errors": 5,
+            "warnings": 10
+          },
+          "cellxgene": {
+            "errors": 2500,
+            "warnings": 20
+          }
+        },
+        "expected_min_length": 249800,
+        "expected_max_length": 250000,
+        "expected_truncated_count": 1
+    },
+    {
+        "name": "multiple_truncated",
+        "description": "Test JSON with multiple message lists requiring truncation",
+        "message_length": 100,
+        "message_counts": {
+          "cap": {
+            "errors": 3500,
+            "warnings": 10
+          },
+          "cellxgene": {
+            "errors": 5,
+            "warnings": 3000
+          }
+        },
+        "expected_min_length": 249800,
+        "expected_max_length": 250000,
+        "expected_truncated_count": 2
+    },
+], ids=lambda x: x["name"])
+def test_length_limited_json_scenarios(test_case):
+  timestamp = "2025-10-18T04:42:04.963Z"
+  list_message = "x" * test_case["message_length"]
+  message = ValidationMessage(
+    file_id="test-file-id",
+    status="success",
+    timestamp=timestamp,
+    bucket="test-bucket",
+    key="test-key",
+    batch_job_id="test-job-id",
+    tool_reports={
+      tool_key: ValidationToolReport(
+        valid=False,
+        errors=[list_message] * tool_counts["errors"],
+        warnings=[list_message] * tool_counts["warnings"],
+        started_at=timestamp,
+        finished_at=timestamp
+      )
+      for tool_key, tool_counts in test_case["message_counts"].items()
+    }
+  )
+  message_json = message.to_length_limited_json()
+  assert test_case["expected_min_length"] <= len(message_json)
+  assert len(message_json) <= test_case["expected_max_length"]
+  assert message_json.count('"Messages truncated"') == test_case["expected_truncated_count"]

